src/osd/PG.cc

```diff
@@ -271,6 +331,7 @@ PG::PG(OSDService *o, OSDMapRef curmap,
 
 PG::~PG()
 {
+  pgstate_history.set_pg_in_destructor();
```

Comment:
I think this destructor-related code could be avoided by declaring pgstate_history after recovery_state in PG.h, so that every NamedState instance is destroyed (and exit() is called) before the pgstate_history object is destroyed.

Moving the pgstate_history instance to PrimaryLogPG might be necessary to handle the snap_trimmer_machine declared there.